### PR TITLE
Document benefits from TypeScript annotation

### LINKED ISSUE #7072: Document how to write plugins using TypeScript
### Describe the documentation issue

I'd like to know how to write and publish plugins using TypeScript.

It seems worthwhile, especially as we:

- publish our types, which can be consumed by plugin authors
- are moving to ESM, which may change how plugins are authored

Ref: https://github.com/stylelint/stylelint/issues/7017#issuecomment-1628823202

### What solution would you like to see?

Improvements to our [writing plugins](https://stylelint.io/developer-guide/rules) documentation, perhaps as a new section,

### PR DESCRIPTION
**Pull request by ybiquitous**
> Which issue, if any, is this issue related to?

Closes #7072

> Is there anything in the PR that needs further explanation?

See https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

### CODE REVIEW

File: docs/developer-guide/plugins.md [hunk 1]
```diff
@@ -34,7 +34,8 @@ const meta = {
   url: "https://github.com/foo-org/stylelint-selector-no-foo/blob/main/README.md"
 };
 
-const ruleFunction = (primary, secondaryOptions) => {
+/** @type {import('stylelint').Rule} */
```
Comment (romainmenke) [inline docs/developer-guide/plugins.md:37]: Should this also demonstrate the generic types that can be declared?

Comment (ybiquitous) [reply docs/developer-guide/plugins.md:37]: Right, the `Rule` type can receive type parameters for primary and secondary options. But I think simplifying the code example is better here so that people can quickly understand it.

Comment (romainmenke) [reply docs/developer-guide/plugins.md:37]: I agree

**Review by romainmenke — APPROVED** [review]
LGTM!

File: docs/developer-guide/plugins.md [hunk 2]
```diff
@@ -66,6 +67,9 @@ ruleFunction.meta = meta;
 export default createPlugin(ruleName, ruleFunction);
 ```
 
+> [!NOTE]
+> The JSDoc annotation (`@type`) benefits auto-completion or type-check by TypeScript.
```
Comment (Mouvedia) [inline docs/developer-guide/plugins.md:71]: minor

There is something wrong with this sentence.
If I had to guess you mean something like
```md
The JSDoc `@type` tag enables auto-completion and type-checking provided by Typescript.
```

Comment (ybiquitous) [reply docs/developer-guide/plugins.md:71]: Looks good. Can you suggest code via **Add a suggestion** button? I'll take your suggestion. 😃 

image

Comment (Mouvedia) [reply docs/developer-guide/plugins.md:71]: I wanted to know if that what you meant first.

```suggestion
> The [JSDoc `@type` tag](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#type) enables the autocompletion and type-checking features provided by Typescript.
```

**Comment by Mouvedia** [comment]
Correct me if I am wrong, but this PR is just one part of #7072.
i.e. it doesn't close it

**Comment by ybiquitous** [comment]
> Correct me if I am wrong, but this PR is just one part of #7072.

I think the current changes are enough, but more suggestions may come. Let's wait.

**Review by jeddy3 — APPROVED** [review]
LGTM, thank you.

File: docs/developer-guide/plugins.md [hunk 3]
```diff
@@ -66,6 +67,9 @@ ruleFunction.meta = meta;
 export default createPlugin(ruleName, ruleFunction);
 ```
 
+> [!NOTE]
+> The [JSDoc `@type` tag](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#type) enables the autocompletion and type-checking features provided by Typescript.
```
Comment (jeddy3) [inline docs/developer-guide/plugins.md:71]:
```suggestion
> The [`@type` JSDoc annotation](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#type) enables Typescript to autocomplete and type-check.
```

File: docs/user-guide/configure.md [hunk 4]
```diff
@@ -14,13 +14,17 @@ Stylelint expects a configuration object, and looks for one in a:
...
+/** @type {import('stylelint').Config} */
 export default {
   rules: {
     "block-no-empty": true
   }
 };
 ```
 
+> [!NOTE]
+> The JSDoc annotation (`@type`) benefits auto-completion or type-check by TypeScript.
```
Comment (jeddy3) [inline docs/user-guide/configure.md:26]:
```suggestion
> The `@type` JSDoc annotation enables Typescript to autocomplete and type-check.
```

**Review by jeddy3 — CHANGES_REQUESTED** [review]
We can reword it to make it simpler.

**Review by jeddy3 — APPROVED** [review]
